Feature/role access

sergeJAVA opened the pull request.

sergeJAVA committed:
feat: добавил jwt для аутентификации, реализую контроллеры, требующие jwt.

sergeJAVA committed:
feat: Добавил контроллеры с эндпоинты, требующие роли для доступа.

sergeJAVA committed:
docs: Added swagger documentation for api's that require authentication.

sergeJAVA committed:
SecurityConfig moved to security package

sergeJAVA committed:
UI controllers moved to ui package

gobaz095 reviewed (commented):
Теперь когда у тебя есть токен, ты можешь получить логин и заполнить служебные поля update_user, create_user

sergeJAVA committed:
feat: modify_user_id и create_user_id теперь заполняются из jwt

Теперь modify_user_id и create_user_id для сущностей Contractor и Deal берутся из jwt

gobaz095 commented on src/main/java/com/internship/deal_service/controller/ui/UIDealController.java:
```diff
@@ -0,0 +1,217 @@
+        TokenAuthentication tokenAuthentication = (TokenAuthentication) authentication;
+        TokenData tokenData = tokenAuthentication.getTokenData();
+
+        List<String> authenticatedUserRoles = tokenData.getAuthorities()
```
Очень громоздко. Лучше вынести в отдельный утильный класс проверки ролей. Можно даже либу сделать. И подтягивать ее в deal и contractor. Туда все роли запихнуть и тд. Чтобы не синхронизировать их руками в каждом сервисе.

gobaz095 commented on src/main/java/com/internship/deal_service/controller/ui/UIContractorRoleController.java:
```diff
@@ -0,0 +1,55 @@
+            @ApiResponse(responseCode = "404", description = "Контрагент или роль не найдены")
+    })
+    @PostMapping("/add")
+    public ResponseEntity<ContractorRoleDto> addRoleToContractor(@RequestBody ContractorRoleRequest request) {
```
https://www.baeldung.com/spring-security-check-user-role

gobaz095 commented on src/main/java/com/internship/deal_service/config/security/SecurityConfig.java:
```diff
@@ -0,0 +1,64 @@
+    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+        return http
+                .authorizeHttpRequests(request ->
+                        request
```
Вместо всего этого лучше анотации, а тут: .requestMatchers("/ui/**").authenticated()

sergeJAVA committed:
feat: перенёс проверку ролей из контроллера в UserRoleUtil;

вместо проверки ролей в filterchain были поставлены аннотации над эндпоинтами.

AlKl1M commented on src/main/java/com/internship/deal_service/util/UserRoleUtil.java:
```diff
@@ -0,0 +1,115 @@
+
+import java.util.List;
+
+public final class UserRoleUtil {
```
Норм, но вообще хорошо было бы это в отдельную библу затолкать - типо у тебя есть сервис аутентификации, отдельно библа с инструментами для аутентификации - какой-нибудь auth-utils-lib - туда конфиги жвт можно засунуть, роли, фильтры и условно этот класс. Это не обязательно все, но в целом хорошая практика т.к. ты этот функционал во многих сервисах использовать будешь.

Если решишь - библы для спринг бута обычно делаются через стартер + автоконфигурацию. 

https://docs.spring.io/spring-boot/docs/2.0.3.RELEASE/reference/html/boot-features-developing-auto-configuration.html#boot-features-custom-starter-naming
https://struchkov.dev/blog/ru/create-spring-boot-starter/
https://docs.spring.io/spring-boot/docs/2.0.3.RELEASE/reference/html/boot-features-developing-auto-configuration.html
https://www.baeldung.com/spring-boot-custom-auto-configuration

Это не обязательно, но для общего развития в целом ознакомиться

AlKl1M reviewed (approved):
По самой задаче у меня замечаний нет, правки глянул. Если Артем что-то заметит - в фиксе подольешь.
